### PR TITLE
Don't use i32 helpers in generated Enum code.

### LINKED ISSUE #356: Don't use Int type helpers in Enum definition
The nimbus Kotlin bindings currently fail to compile with:

```
[task 2020-12-08T06:07:27.105Z] > Task :nimbus:compileDebugKotlin FAILED
[task 2020-12-08T06:07:27.105Z] e: /builds/worker/checkouts/src/components/nimbus/build/generated/source/uniffi/debug/java/org/mozilla/experiments/nimbus/nimbus.kt: (606, 55): Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
[task 2020-12-08T06:07:27.105Z] internal fun kotlin.Boolean.Companion.read(buf: ByteBuffer): Boolean defined in org.mozilla.experiments.nimbus in file nimbus.kt
[task 2020-12-08T06:07:27.105Z] internal fun kotlin.Byte.Companion.read(buf: ByteBuffer): Byte defined in org.mozilla.experiments.nimbus in file nimbus.kt
[task 2020-12-08T06:07:27.105Z] internal fun kotlin.String.Companion.read(buf: ByteBuffer): String defined in org.mozilla.experiments.nimbus in file nimbus.kt
```

The line causing the error is the `read` definition of an `enum class`, which is attempting to use `Int.read` as a helper function. I believe this is failing because the nimbus IDL doesn't use `i32` anywhere, so we don't generate any of the helpers for `Int`.

The generated code should use `buf.getInt()` directly, rather than the `Int.read` helper. PR incoming.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-1926)

### PR DESCRIPTION
**Pull request by rfk**
Previously, the generated Kotlin code for enums would try to use
helpers from the `Int` companion class, but we would only emit the
code for those helpers if the interface was also using the integer
type.

Now, the generated Kotlin code for enums directly uses lower-level
methods for reading and writing integers, so that it doesn't depend
on other generated code.

Fixes #356.

### CODE REVIEW

**Comment by rfk** [comment]
@jhugman r? This blocks the release of an updated nimbus over in https://github.com/mozilla/application-services/pull/3748.

File: Cargo.toml
Comment (rfk) [inline Cargo.toml:12]: I didn't want to land this without a test, and I didn't want to contrive a new example just to test this edge-case (especially since the bug depends on the *absence* of something in the UDL, rather than its presence). So, I'm trying out a new suite of explicit regression tests.

The setup here is a little wordy for my liking, and if we generate more of these then we may want to invest in some automation so they can be written more succinctly. But it's a start!

Comment (jhugman) [reply Cargo.toml:12]: I'm so pleased you have a test!

I think we could bike shed this, but happy to land as is.

FWIW, I think my pref would be:

`fixtures/regressions/enums-without-i32-helper` and then put the URL to the issue in the README.md of the fixture.

File: regression_tests/README.md
Comment (jhugman) [inline regression_tests/README.md:6]: Nit: uniffi

File: regression_tests/i356-enum-without-int-helpers/src/test.udl
Comment (jhugman) [inline regression_tests/i356-enum-without-int-helpers/src/test.udl:3]: Nice.

File: regression_tests/i356-enum-without-int-helpers/tests/bindings/test.swift
Comment (jhugman) [inline regression_tests/i356-enum-without-int-helpers/tests/bindings/test.swift:4]: Nit: exclude `Which`. i.e. 

```swift
assert(which(arg: true) == .yeah)
```

Comment (jhugman) [reply regression_tests/i356-enum-without-int-helpers/tests/bindings/test.swift:4]: Also: I note that Swift is using the `UInt32.read` extension methods, so should we change that now? or file an issue?

File: regression_tests/i356-enum-without-int-helpers/tests/bindings/test.kts
Comment (jhugman) [inline regression_tests/i356-enum-without-int-helpers/tests/bindings/test.kts:4]: So I guess this wouldn't work before, but now it does? That's what makes it a useful test?

**Review by jhugman — APPROVED** [review]
I have a few nits, but nothing that needs a second round of review.

File: regression_tests/i356-enum-without-int-helpers/tests/bindings/test.kts
Comment (rfk) [reply regression_tests/i356-enum-without-int-helpers/tests/bindings/test.kts:4]: Prior to the fix, we wouldn't even get to running this, it would fail to compile. Which, it's probably worth me adding a comment to that effect in the test here for clarity.

File: Cargo.toml
Comment (rfk) [reply Cargo.toml:12]: WFM, I will update to the suggested layout. Something about the shorter directory names is very pleasing :-)

File: regression_tests/i356-enum-without-int-helpers/tests/bindings/test.swift
Comment (rfk) [reply regression_tests/i356-enum-without-int-helpers/tests/bindings/test.swift:4]: Nice catch. It looks like the swift bindings generate these helpers unconditionally, rather than inspecting the types used in the UDL. Let's file an issue for followup to do conditional generation, and we can count on this new test to catch the problem if we reintroduce it there. (I confirmed that the test would catch it by commenting out the `extension UInt32: Primitive, ViaFfi` in the generated swift bindings, and this swift test does indeed fail with an appropriate compile-time error in that case).

Comment (rfk) [reply regression_tests/i356-enum-without-int-helpers/tests/bindings/test.swift:4]: I filed https://github.com/mozilla/uniffi-rs/issues/362 for followup.

**Comment by rfk** [comment]
Updated per review feedback, I'll merge this once CI goes green.

**Pull request merged**
